demos/signature/rsa_pss_direct.c

```diff
@@ -0,0 +1,186 @@
+    if (pkey == NULL) {
+        fprintf(stderr, "Failed to load private key\n");
+        goto end;
+    }
+
+    /* Create signing context. */
+    ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, propq);
+    if (ctx == NULL) {
+        fprintf(stderr, "Failed to create signing context\n");
+        goto end;
+    }
+
+    /* Initialize context for signing and set options. */
+    if (EVP_PKEY_sign_init(ctx) == 0) {
+        fprintf(stderr, "Failed to initialize signing context\n");
+        goto end;
+    }
+
+    if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING) == 0) {
+        fprintf(stderr, "Failed to configure padding\n");
+        goto end;
+    }
+
+    if (EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()) == 0) {
+        fprintf(stderr, "Failed to configure digest type\n");
+        goto end;
+    }
+
+    /* Determine length of signature. */
+    if (EVP_PKEY_sign(ctx, NULL, sig_len,
+                      test_digest, sizeof(test_digest)) == 0) {
+        fprintf(stderr, "Failed to get signature length\n");
+        goto end;
+    }
+
+    /* Allocate memory for signature. */
+    *sig = OPENSSL_malloc(*sig_len);
+    if (*sig == NULL) {
+        fprintf(stderr, "Failed to allocate memory for signature\n");
+        goto end;
+    }
+
+    /* Generate signature. */
+    if (EVP_PKEY_sign(ctx, *sig, sig_len,
+                      test_digest, sizeof(test_digest)) != 1) {
+        fprintf(stderr, "Failed to sign\n");
+        goto end;
+    }
+
+    rv = 1;
+end:
+    EVP_PKEY_CTX_free(ctx);
+    EVP_PKEY_free(pkey);
+
+    if (rv == 0)
+        OPENSSL_free(*sig);
+
+    OSSL_LIB_CTX_free(libctx);
+    return rv;
+}
+
+/*
+ * This function demonstrates verification of an RSA signature over a SHA-256
+ * digest using the PSS signature scheme.
+ */
+static int verify(const unsigned char *sig, size_t sig_len)
+{
+    int rv = 0;
+    OSSL_LIB_CTX *libctx = NULL;
+    const unsigned char *ppub_key = NULL;
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+
+    /* Load DER-encoded RSA public key. */
+    ppub_key = rsa_pub_key;
+    pkey = d2i_PublicKey(EVP_PKEY_RSA, NULL, &ppub_key, sizeof(rsa_pub_key));
+    if (pkey == NULL) {
+        fprintf(stderr, "Failed to load public key\n");
+        goto end;
+    }
+
+    /* Create verification context. */
+    ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, propq);
+    if (ctx == NULL) {
+        fprintf(stderr, "Failed to create verification context\n");
+        goto end;
+    }
+
+    /* Initialize context for verification and set options. */
+    if (EVP_PKEY_verify_init(ctx) == 0) {
+        fprintf(stderr, "Failed to initialize verification context\n");
+        goto end;
+    }
+
+    if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING) == 0) {
+        fprintf(stderr, "Failed to configure padding\n");
+        goto end;
+    }
+
+    if (EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()) == 0) {
```

Comment:
I'd have preferred to see EVP_MD_fetch() rather than the older EVP_sha256().
It's not all that important.

A future possibility would, perhaps, be an EVP_PKEY_CTX_set_signature_md_ex() that takes a name and property query string.

Reply:
This is worth fixing. Pushed.

Reply:
THanks.